Allow materialized views to by synchronous

havaker opened the pull request:
This pull request introduces a "synchronous mode" for global views. In this mode, all view updates are applied synchronously as if the view was local.

Marking view as a synchronous one can be done using `CREATE MATERIALIZED VIEW` and `ALTER MATERIALIZED VIEW`. E.g.:
```cql
ALTER MATERIALIZED VIEW ks.v WITH synchronous_updates = true;
```

Marking view as a synchronous one was done using tags (originally used by alternator). No big modifications in the view's code were needed.

Fixes: https://github.com/scylladb/scylla/issues/10545

nyh commented:
If this PR does what issue #10545 requests to do (doesn't it?), it should say "Fixes #10545", not Refs.

nyh commented on db/tags/utils.hh:
```diff
@@ -0,0 +1,39 @@
+
+// get_tags_of_table() returns all tags associated with the given table, or
+// nothing if the table is missing the tags extension.
+std::optional<std::reference_wrapper<const std::map<sstring, sstring>>> get_tags_of_table(schema_ptr schema);
```
Now that this code is out of Alternator it becomes even more important to accurately document this API.  I would ask myself what does the returned reference point to, and how long is it guaranteed to be alive.

By the way, instead of returning an optional<reference>, isn't the more classic approach to use a **pointer**, which is more-or-less an optional reference?

psarna replied:
We do have quite a few precedents of using a pointer like that in the code, so feel free to use it as well. It's theoretically less safe, but on the other hand it's quite obvious that we need to do a null check on a pointer returned by some helper function, so I consider it "safe enough"

nyh reviewed (commented):
Looks good to me, except that I'm not sure if using the Alternator tags technique is a good idea, or a bad idea. It's bad in the sense that we already another dozen table options so why do we want this one to be different, but good in the sense that I also used it in Alternator TTL to configure the expiration column, so if one day we want to support that TTL implementation also outside Alternator, it will be useful to have the tags support also outside Alternator.
@psarna do you have any thoughts about this?

By the way, I was thinking - if we add a generic "tags" feature in the implementation, could we - or should we - also add to "ALTER/CREATE TABLE" the ability to write any tag, and not just one specific one? On the other hand, maybe we should check the "system:*" tags if they are valid - otherwise a user can have a spelling mistake and not notice... So I don't know if this is a good idea or not.

psarna commented:
@nyh I suggested moving tags out of `alternator` and reusing them for this purpose, as an alternative to adding a column to one of the existing system tables, which brings backward compatibility issues. Tags sounded like a natural solution here, because we indeed just wanted to tag specific views in order to make them special.

nyh commented on db/tags/extension.hh:
```diff
@@ -37,4 +37,9 @@ private:
+// Information whether the view updates are synchronous is stored using the
+// SYNCHRONOUS_VIEW_UPDATES_TAG_KEY tag. Value of this tag is a stored as a
+// serialized boolean value ("true" or "false")
+static const sstring SYNCHRONOUS_VIEW_UPDATES_TAG_KEY("system:synchronous_view_updates");
```
Currently in in Alternator, you can't tag a view table, just the base table...
https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Tagging.html says "Any local secondary indexes (LSI) and global secondary indexes (GSI) related to tagged tables are labeled with the same tags automatically." but we haven't implemented that. We could, I guess.

Thinking about this made me ask myself a question: do we really want "synchronous updates" to be based on a tag on the view table? Wouldn't it be even more convenient to put a tag on the base table, and it will apply to writes to all its views? Or do we envision a situation where a user might want to make some views synchronous and some not?

psarna replied:
When I thought about https://github.com/scylladb/scylla/issues/10545 I thought that this granularity may be useful. I imagined a case when somebody wants to create X materialized views that are just "best effort" and can be rebuilt or even dropped if necessary, but they also care about consistency for a single specific view. Then, only this view is marked as synchronous and we only "pay" with availability for this one, while the rest is still asynchronous and does not cause the base write to fail. Since synchronous updates sent over the network are costly, I figured it's better to have higher granularity, especially that the implementation is not particularly complicated.

nyh replied:
I see. It makes sense. We could also allow *both* options (if you tag the base table, it applies to all views, or you can tag only an individual view) but that can get complicated.
As long as we can only tag the view, we should be aware that doing this is not currently possible in Alternator. I'll open an issue.

nyh replied:
I opened https://github.com/scylladb/scylla/issues/10786 about the tagging of materialized views in Alternator (which should be allowed in one form or another). But this doesn't block this patch - we can start supporting this feature before Alternator does.

havaker commented:
v2:

- changed the signature of `get_tags_of_table` to use a pointer instead of `std::optional<std::reference_wrapper<const T>`
- renamed `should_wait_for_all` to `apply_updates_synchronously` and made it `const`
- added requested trace point
- improved messages in the throw statements
- added lifetime info about the value returned by `get_tags_of_table`

nyh commented on db/tags/utils.cc:
```diff
@@ -0,0 +1,56 @@
+
+namespace db {
+
+const std::map<sstring, sstring>* get_tags_of_table(schema_ptr schema) {
```
The commit message still refers to std::optional<std::ref>, please update it now that you have a pointer instead.

nyh commented on db/view/view.cc:
```diff
@@ -1226,6 +1237,18 @@ future<> mutate_MV(
+            }
+
+            return wait_for_all || update_synchronously;
+        } ();
```
Why is this a lambda called immediately, instead of just staightforward code which ends with something like

const bool apply_update_sychronously = wait_for_all || update_synchronously

A lambda is useful when you call it multiple times. Here you only call it once... I also got confused by this initially because I thought the variable held the lambda, while it holds the result of it.

havaker replied:
I had the impression that a lot of similar sounding variables (`apply_update_synchronously`, `update_synchronously`) in the same scope could cause confusion, so I hid the `update_synchronously` variable in its own block. I agree that instantly called lambda is not the cleanest solution, I will change it to what you proposed.

nyh commented:
@havaker please clarify on the commit message if this Fixes or just Refs the issue, and also move it out of "draft" mode when you think the patch is ready.
How did you test this code, or thinking of testing it?

havaker commented:
> If this PR does what issue #10545 requests to do (doesn't it?), it should say "Fixes #10545", not Refs.

Right, it should.

I wasn't so sure if it would work with indexes - that's why I left "Refs" at first. After learning a bit more about them (mainly that they are represented by views), I can safely mark it as "Fixes".

havaker commented:
> How did you test this code, or thinking of testing it?

I haven't yet. My idea was to move it out of "draft" mode after writing a dtest aiming at checking this functionality.

havaker commented:
v3:
- updated commit message to mention pointer instead of optional
- removed the instantly called lambda

psarna reviewed (changes requested):
Looks good, left a few nitpicks and I think the commit message needs rephrasing

havaker commented on cql3/statements/cf_prop_defs.cc:
```diff
@@ -77,7 +78,8 @@ void cf_prop_defs::validate(const data_dictionary::database db, sstring ks_name,
         KW_BF_FP_CHANCE, KW_MEMTABLE_FLUSH_PERIOD, KW_COMPACTION,
-        KW_COMPRESSION, KW_CRC_CHECK_CHANCE, KW_ID, KW_PAXOSGRACESECONDS
+        KW_COMPRESSION, KW_CRC_CHECK_CHANCE, KW_SYNCHRONOUS_UPDATES, KW_ID,
+        KW_PAXOSGRACESECONDS
```
I don't remember the exact reason, I will move it to the end.

nyh commented on cql3/statements/alter_table_statement.cc:
```diff
@@ -341,6 +341,10 @@ std::pair<schema_builder, std::vector<view_ptr>> alter_table_statement::prepare_
                 }
             }
 
+            if (_properties->get_synchronous_updates_flag()) {
```
As I noted in https://github.com/scylladb/scylla/issues/10545#issuecomment-1123920515, it would be nice to try to understand and document somewhere (preferably not a commit message that nobody will ever see again. Maybe docs/?) exactly what are the real advantages and disadvantages of this option. We discussed on that thread some non-options.

That being said, the **feature** can be committed even before we have good documentation for it.

psarna replied:
> Are you going to change it as well?

I will, definitely, it's the same kind of oversimplification; thanks

havaker commented:
v4:

- nitpicks from @psarna 
- rephrased [cql3: statements: add a synchronous_updates property](https://github.com/scylladb/scylla/pull/10777/commits/8eafd9562ccdbe81aba686d5d4b7b63ef7411623) commit message
- rebase

havaker commented:
v5:
- tests
- rebase

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/999/

psarna commented:
I'm going to rebase this series and move it to my own branch, since I don't have write rights for the original branch and we should probably try to move this forward before it bitrots

psarna commented:
Superseded by #11013
